[IMP] website, website_mass_mailing: add new `s_newsletter_aside` snippet

anso-odoo opened the pull request:
This PR adds the new `s_newsletter_aside` snippet.

It also adds an option in the form editor to allow the standard checkbox layout: "checkbox + label" that was needed to achieve the desired design.

task-4306896

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/188802.png)](https://mergebot.odoo.com/odoo/odoo/pull/188802)

stefanorigano commented on addons/website/static/src/xml/website_form_editor.xml:
```diff
@@ -39,6 +39,15 @@
                 </div>
             </div>
+            <t t-elif="field.formatInfo.labelPosition == 'after'">
+                <div class="d-flex">
```
Rather than `flex`, I think that we should use a `.checkbox > .form-check` structure here 🤔
You can see it implemented already for the "_Multiple Checkboxes_" field type.
ref: https://getbootstrap.com/docs/5.3/forms/checks-radios/#checks

stefanorigano commented on addons/website/static/src/xml/website_form_editor.xml:
```diff
@@ -39,6 +39,15 @@
+            <t t-elif="field.formatInfo.labelPosition == 'after'">
+                <div class="d-flex">
+                    <t t-out="0"/>
+                    <label class="s_website_form_label s_website_form_label_after" t-attf-style="width: #{field.formatInfo.labelWidth or '200px'}" t-att-for="field.id">
```
` s_website_form_label_after ` :  unless I’m missing something, it looks like we’re creating a new class just to set its corresponding UI button to "active"...

Read my previous comment, you may be able to use `form-check-label` instead 😉

stefanorigano commented on addons/website/static/src/snippets/s_website_form/options.js:
```diff
@@ -343,8 +343,16 @@ const FieldEditor = FormEditor.extend({
     _getLabelPosition: function () {
         const label = this.$target[0].querySelector('.s_website_form_label');
         if (this.$target[0].querySelector('.row:not(.s_website_form_multiple)')) {
-            return label.classList.contains('text-end') ? 'right' : 'left';
+            if (label.classList.contains('text-end')) {
+                return 'right';
+            } else if (label.classList.contains('s_website_form_label_after')) {
+                return 'after';
+            }
+            return 'left';
         } else {
+            if (label.classList.contains('s_website_form_label_after')) {
+                return 'after';
+            }
             return label.classList.contains('d-none') ? 'none' : 'top';
```
We may simplify this a bit 😉

```js
_getLabelPosition: function () {
    const label = this.$target[0].querySelector('.s_website_form_label');
    const isRow = this.$target[0].querySelector('.row:not(.s_website_form_multiple)');

    if (label.classList.contains('form-check-label')) {
        return 'after';
    }
    if (isRow) {
        return label.classList.contains('text-end') ? 'right' : 'left';
    }
    return label.classList.contains('d-none') ? 'none' : 'top';
},
```
note: to be tested. It also assumes that we can use `form-check-label` class.

stefanorigano reviewed (changes requested):
Hello @anso-odoo , here some comments about the new [checkbox label option](70a1d9895eb2e67521cb1d6a0f7b7c3a0955bef4) (I didn't check the snippet itself yet).
Visual result it's ok, my remarks are just about code optimization and consistency.

Since we're here, I think we should also disable the option to hide the checkbox labels, it doesn't make sense to keep it IMO.
Adding @lebl-odoo in the loop, I think we need his insights here.

Thanks!

lebl-odoo commented:
Actually, if we implement a double opt-in, we should be able to remove the checkbox since visitors should validate their email address first. It is good from a GDPR point view. Let's move in that direction. 

Create an option for single opt-in. By default, it is not shown. 
![image](https://github.com/user-attachments/assets/933ca289-4562-4021-9a62-37bcd33af7f4)

Another though: should we display `list_ids` field if we propose only one choice?

(Note that I didn't test yet since the task is still in development)

lebl-odoo commented:
> I don't know if it changes something in our side tho, or are you proposing to hide the block if the user is already logged-in?

A logged in users (e.g. a customer) could fill in the newsletter form to be contacted. I had a look, for the newsletter popup, it is displayed when we're logged in (as admin).

> About the change anso proposed; it's about adding an option to change the checkbox's label position, I think that it makes sense regardless by the newsletter snippets, wdyt? Also, what do you think about removing the option to hide the checkboxes' labels?

Not sure why, but I cannot access to the [commit](https://github.com/odoo/odoo/pull/70a1d9895eb2e67521cb1d6a0f7b7c3a0955bef4) you mentionned. I don't have the context. Do you mean to have the checkbox left or right to the label?

> Another though: should we display list_ids field if we propose only one choice?

![image](https://github.com/user-attachments/assets/20282820-c565-4fd9-a211-3bbb978486b9)

If we display only one newsletter to subscribe, I don't think we should display the checkbox.

anso-odoo commented:
Hello @lebl-odoo 
As discussed in private, there is not difference in what we show to a connected or a not connected user so we are good on this.

> Not sure why, but I cannot access to the [commit](https://github.com/odoo/odoo/pull/70a1d9895eb2e67521cb1d6a0f7b7c3a0955bef4) you mentionned. I don't have the context. Do you mean to have the checkbox left or right to the label?

Concerning this, the commit is [this one ](https://github.com/odoo/odoo/pull/188802/commits/3c24dd6bdf7878580f9c2f6293523efee78ad3c3). Basically what we do is to add an option in the editor to be able to achieve the default layout of a checkbox, what is not possible to do now :

![Group 84](https://github.com/user-attachments/assets/994dee65-343a-4e87-aec8-eebac63a0f87)
<img width="261" alt="Screenshot 2024-12-02 at 10 27 14" src="https://github.com/user-attachments/assets/d82ebaeb-80e7-49fc-86d9-09cace0805fa">

> If we display only one newsletter to subscribe, I don't think we should display the checkbox.

I see why you are saying this but then, shouldn't we just remove the "I agree to receive updates" checkbox in any case ?

1. This field is just a required checkbox placed there for RGPD I guess, but we can totally live without it. We could add a description to the mailing lists field that would do the same RGPD job <img width="734" alt="Screenshot 2024-12-02 at 11 27 46" src="https://github.com/user-attachments/assets/ca93cb5c-8933-43b5-9f2b-637f9ccbde84">
2. If there is no mailing list set in the backend, the newsletter snippet is not usable anyway (I receive this error when I try to drop it)<img width="1170" alt="Screenshot 2024-12-02 at 11 17 37" src="https://github.com/user-attachments/assets/2ee5ff41-7f89-4cd7-a638-4745ef535321">
3. If I have multiple mailing lists, I'm forced to check the mailing list to fill the form. The "I agree to receive updates" just seems like a redundancy and is not very clear.<img width="663" alt="Screenshot 2024-12-02 at 11 22 10" src="https://github.com/user-attachments/assets/178ed1ac-6efb-456f-8cfa-39915fd91603">

I'm not sure it's in the scope of this task though. Because if we do it here, we should do the same change for the other newsletter form templates.

lebl-odoo commented:
> This field is just a required checkbox placed there for RGPD I guess, but we can totally live without it. We could add a description to the mailing lists field that would do the same RGPD job.
Currently, I can check and add your email address, but you never consent. This is why I proposed to create an option for single opt-in. By default, it is not shown.

> Basically what we do is to add an option in the editor to be able to achieve the default layout of a checkbox
Nice idea, we worked on a [equivalent task](https://www.odoo.com/odoo/project/974/tasks/3922625) but I just paused it since your behavior seems more intuitive

> Bonus point: If there is no mailing list set in the backend, the newsletter snippet is not usable (I receive this error when I try to drop it)
Should the website team take care of this or are you able to take into account?

See you wednesday to clarify the rest :)

anso-odoo commented:
@lebl-odoo 

As we discussed:
1. I removed the useless solo "RGPD" checkbox to keep only the mailing lists ones 
2. I removed the checkbox alignment commit as it will be handle in [this task](https://www.odoo.com/odoo/project/974/tasks/3922625) or in another website forms refinement task (it was not useful anymore in this PR anyway, thanks to "1.")
3. Now, when you have only one `mailing_list` visible in the font-end, there is no checkbox
4. Fixed the error when there was no `mailing_list` set in the backend

Can you check? Thanks!

lebl-odoo commented:
Hi there :wave: 

I had another look and here is what I found:

A small layout issue, de Display option shouldn't be displayed as a suboption
![image](https://github.com/user-attachments/assets/86f6b7f0-ff4e-48c2-9200-61cc4facb57c)

When I select only one newsletter, I expect to hide the checkbox in the frontend (behavior: if I submit my form, I subscribed to the hidden newsletter).

Brieuc-brd commented on addons/website_mass_mailing/views/snippets_templates.xml:
```diff
@@ -290,6 +295,100 @@ database, without the s_newsletter_list class. See fixNewsletterListClass.
+                                    </label>
+                                    <input id="mailing_sub1" type="text" class="form-control s_website_form_input" name="name" required="1" placeholder="John Smith"/>
+                                </div>
+                                <div t-attf-class="#{'col-lg-12' if request.env['mailing.contact']._is_name_split_activated() else 'col-lg-6'} mb-3 s_website_form_field s_website_form_model_required" data-type="email" data-name="Field">
```
Are you sure this condition works ? 🤔 
The design doesn't change even if I change this option in the settings, am I missing something ? 😅 
<img width="1244" height="456" alt="Capture d’écran 2025-10-02 à 11 08 29" src="https://github.com/user-attachments/assets/b2ad77d3-d697-4c56-b2e0-cedd22624d16" />

anso-odoo replied:
If you have already the snippet in the page, you'll need to add a new one to have the option active...

Not the best I agree but it's already like that on master 🤔

Brieuc-brd commented on addons/website_mass_mailing/views/snippets_templates.xml:
```diff
@@ -290,6 +295,100 @@ database, without the s_newsletter_list class. See fixNewsletterListClass.
+                                </t>
+                                <div class="s_website_form_submit s_website_form_no_submit_label col-12" data-name="Submit Button">
+                                    <div style="width: 250px;" class="s_website_form_label"/>
+                                    <span id="s_website_form_result"></span>
```
```suggestion
                                    <span id="s_website_form_result"/>
```

Also, I think it should be placed after the button, so the error message won't push the submit button :
<img width="784" height="237" alt="Capture d’écran 2025-10-02 à 11 22 53" src="https://github.com/user-attachments/assets/2f59f3e5-0755-4b63-8698-47bf4a61cd52" />

Brieuc-brd commented on addons/website_mass_mailing/views/snippets_templates.xml:
```diff
@@ -290,6 +295,100 @@ database, without the s_newsletter_list class. See fixNewsletterListClass.
+                                    </div>
+                                </t>
+                                <div class="s_website_form_submit s_website_form_no_submit_label col-12" data-name="Submit Button">
+                                    <div style="width: 250px;" class="s_website_form_label"/>
```
Is this empty div really necessary ? 🤔

Brieuc-brd reviewed (changes requested):
Thanks for your work @anso-odoo 👍 
I've left some feedbacks below.

Also, not sure it's related to your PR but some other Contact snippets seem broken in your branch, maybe a rebase would fix them ? Can you check ?

| | | |
|--------|--------|--------|
| <img width="1152" height="802" alt="Capture d’écran 2025-10-02 à 11 26 10" src="https://github.com/user-attachments/assets/c4a18662-2da2-4603-93a9-a72b71415c81" /> | <img width="1488" height="658" alt="Capture d’écran 2025-10-02 à 11 26 38" src="https://github.com/user-attachments/assets/4933db16-ccc0-4ea2-87c7-9268b0b72443" /> | <img width="1487" height="642" alt="Capture d’écran 2025-10-02 à 11 26 46" src="https://github.com/user-attachments/assets/e4fde521-408b-44d9-a5d6-c5eb68703e06" /> |

anso-odoo commented on addons/website_mass_mailing/views/snippets_templates.xml:
```diff
@@ -290,6 +295,100 @@ database, without the s_newsletter_list class. See fixNewsletterListClass.
+                        <form id="newsletter_form" action="/website/form/" method="post" enctype="multipart/form-data" class="o_mark_required"
+                                data-mark="*" data-model_name="mailing.contact" data-success-mode="message" hide-change-model="true">
+                            <div class="s_website_form_rows row s_col_no_bgcolor">
+                                <div t-if="request.env['mailing.contact']._is_name_split_activated()" class="s_website_form_field s_website_form_model_required col-lg-6 mb-3" data-type="char" data-name="Field">
```
Actually I'm using the default layout of `website.form_field` template in `website_form_editor.xml`

It seems like the classes of `.s_website_form_field` are reset with JS, so if I apply `pb16` it will be erased and the `mb-3` will still be visible

anso-odoo commented:
@Brieuc-brd Thanks, I rebased and adapted the PR. Tell me what you think of the comments when you have time

Brieuc-brd reviewed (changes requested):
Nice, just a small comment below 👍 

Also, it seems that labels of checkbox list have disappeared, could you check ?
<img width="669" height="234" alt="Capture d’écran 2025-10-06 à 08 55 25" src="https://github.com/user-attachments/assets/8deb355e-cd4a-4bd0-88b3-c445596caa41" />

anso-odoo commented on addons/website_mass_mailing/views/snippets_templates.xml:
```diff
@@ -290,6 +295,99 @@ database, without the s_newsletter_list class. See fixNewsletterListClass.
+                                <t groups="website.group_website_designer">
+                                    <t t-set="mailing_lists" t-value="request.env['mailing.list'].search([('is_public', '=', True)])"/>
+                                    <div t-attf-class="#{'' if len(mailing_lists) > 1 else 'd-none'} col-lg-12 s_website_form_field s_website_form_model_required pb24" t-if="len(mailing_lists) > 0" data-type="many2many" data-name="Field">
+                                        <label class="col-sm-auto s_website_form_label d-none" style="width: 250px" for="mailing_list_">
```
Actually I kept this part from the existing template (`.s_newsletter_block_form_template`) but it doesn't seem useful, so I'll just remove that

anso-odoo commented:
@Brieuc-brd Thanks, labels are back, everything should be fine now 👍

Brieuc-brd reviewed (approved):
LGTM, Good job 👍
Note: As you probably already know, the Enterprise test is 🔴 . I'm not sure why, I leave that to the technical review 🙂 .
